sdk/core/core-http/src/policies/bearerTokenAuthenticationPolicy.ts

```diff
@@ -10,107 +10,210 @@ import {
 import { Constants } from "../util/constants";
 import { HttpOperationResponse } from "../httpOperationResponse";
-import { HttpHeaders } from "../httpHeaders";
 import { WebResourceLike } from "../webResource";
-import { AccessTokenCache, ExpiringAccessTokenCache } from "../credentials/accessTokenCache";
-import { AccessTokenRefresher } from "../credentials/accessTokenRefresher";
+import { delay } from "../util/utils";
+
+// #region Access Token Cycler
 
 /**
- * The automated token refresh will only start to happen at the
- * expiration date minus the value of timeBetweenRefreshAttemptsInMs,
- * which is by default 30 seconds.
+ * A function that gets a promise of an access token and allows providing
+ * options.
+ *
+ * @param options - the options to pass to the underlying token provider
  */
-const timeBetweenRefreshAttemptsInMs = 30000;
+type AccessTokenGetter = (options: GetTokenOptions) => Promise<AccessToken>;
+
+interface TokenCyclerOptions {
+  /**
+   * The window of time before token expiration during which the token will be
+   * considered unusable due to risk of the token expiring before sending the
+   * request.
+   *
+   * This will only become meaningful if the refresh fails for over
+   * (refreshWindow - forcedRefreshWindow) milliseconds.
+   */
+  forcedRefreshWindow: number;
+  /**
+   * Interval in milliseconds to retry failed token refreshes.
+   */
+  retryInterval: number;
+  /**
+   * The window of time before token expiration during which
+   * we will attempt to refresh the token.
+   */
+  refreshWindow: number;
+}
+
+// Default options for the cycler if none are provided
+export const DEFAULT_CYCLER_OPTIONS: TokenCyclerOptions = {
+  forcedRefreshWindow: 1000, // Force waiting for a refresh 1s before the token expires
+  retryInterval: 3000, // Allow refresh attempts every 3s
+  refreshWindow: 1000 * 60 * 2 // Start refreshing 2m before expiry
+};
 
 /**
- * Creates a new BearerTokenAuthenticationPolicy factory.
+ * Converts an an unreliable access token getter (which may resolve with null)
+ * into an AccessTokenGetter by retrying the unreliable getter in a regular
+ * interval.
  *
- * @param credential - The TokenCredential implementation that can supply the bearer token.
- * @param scopes - The scopes for which the bearer token applies.
+ * @param tryGetAccessToken - a function that produces a promise of an access
+ * token that may fail by returning null
+ * @param retryInterval - the time (in milliseconds) to wait between retry
+ * attempts
+ * @returns - a promise that, if it resolves, will resolve with an access token
  */
-export function bearerTokenAuthenticationPolicy(
-  credential: TokenCredential,
-  scopes: string | string[]
-): RequestPolicyFactory {
-  const tokenCache: AccessTokenCache = new ExpiringAccessTokenCache();
-  const tokenRefresher = new AccessTokenRefresher(
-    credential,
-    scopes,
-    timeBetweenRefreshAttemptsInMs
-  );
+async function beginRefresh(
```

Comment:
so this never gives up? Is there any case it could get stuck forever? What happens if I pass in a token credential with a bad endpoint or something?

Reply:
I can't think of any reason a TokenCredential would actually return null. It bothers me that the API is written in such a way that resolving with null is possible, but I think for compatibility reasons we are stuck with it. What I guess we should be doing here instead of checking for null is catching rejections.

We could just use a timer? If the refresh doesn't complete within five seconds after expiration (or within five seconds period if there is no token), we can just reject the refresher. Any requests queued up waiting for that refresh job will reject as well.

Reply:
I added the logic proposed by @sadasant below. If the existing token has already timed out, we retry one more time and then surface an exception to all awaiters if it failed again.